full_snapshot.proto

```diff
@@ -291,6 +291,7 @@ message RelationInformation {
   bool has_parent_relation = 16;
 
   string partition_boundary = 17;
+  repeated int32 partition_columns = 18; // For tables range-partitioned by one or more columns, the list of columns
```

Comment:
It seems that partition columns could also be expressions - should we support that? (see https://www.postgresql.org/docs/12/catalog-pg-partitioned-table.html)

Separately, we should also start tracking the partitioning strategy (`pg_partitioned_table.partstrat`)

Reply:
Should we? I felt starting with only supporting partstrat `r` and ignoring any expression-based partitioning lets us land the most common case pretty quickly, but we can certainly add the other modes. What's the most useful way to represent `partexprs` in the snapshot if we want to support that?

Reply:
This is how I filter for that in the collector patch right now: https://github.com/pganalyze/collector/pull/65/files#diff-73bc6d76f51a62d4cbadf2a7899b09bfR20 (but down to do full partstrat support if you think that's useful or will save pain down the line).

Reply:
As per discussion, I'll review how common other strategies and function-based partitioning is and proceed accordingly.

Reply:
So the Postgres docs mostly talk about range-based partitioning and barely mention the other methods: https://www.postgresql.org/docs/current/ddl-partitioning.html -- that said, it looks like supporting list-based and hash-based partitioning is pretty trivial, so we should definitely add those (I'd only need to change the collection query). For expression-based partitions, that's fairly easy to collect as well (we can use `pg_get_expr`), but we'd need to figure out how to represent it in the snapshot state. It's essentially a union: instead of a column reference `PartitionColumns` field like I added, we'd need to do something like

```go
type PartitionComponent {
  Column int32
  Expr null.String
}
```

and either have Column set to -1 or Expr set to null (we could also add a Kind field or IsExpr boolean flag to be more explicit). It's probably worth doing just to not have to worry about adding it later, just not sure about the structure--I don't see anything similar in the state package to go off of.

Also, we can add OpClass field to that struct if we want to cover everything, though that seems pretty rare.

Thoughts?

Reply:
That sounds good, i.e. introduce a field for the partition type + change the columns to be partition components instead (so we support expressions)

For the partition component we could use oneof: https://developers.google.com/protocol-buffers/docs/proto#oneof

For the partition type we could use enum: https://developers.google.com/protocol-buffers/docs/proto#enum

Also could use `PartitionField` instead of `PartitionComponent` as a name

Reply:
@lfittl so this is trickier than I thought because of how it's encoded in the `pg_partitioned_table`:
 * `partattrs` is an int2vector of column numbers for any partition field that is a column or zero for any partition field that is an expression
 * `partexprs` is a single `pg_node_tree` of all expressions used in any fields

As far as I can tell, we can use `pg_get_expr` on `partexprs`, but that gives us a single string of comma-separated expressions, and no easy way to break them up (since expressions can contain commas). I don't see any functions that operate on `pg_node_tree` that could give us the individual elements.

I looked at what psql is using for `\d` and turns out there's actually a `pg_get_partkeydef`:

```
maciek=# SELECT pg_catalog.pg_get_partkeydef('4061537'::pg_catalog.oid);
           pg_get_partkeydef           
---------------------------------------
 RANGE (((a + 2)), c, substr(b, 2, 1))
(1 row)
```

We could use a `partitioned_by` string field and just populate with this function. If we want to track first-class information for the common case of columns and/or partitioning strategy, we could also add those separately, but as far as I can see, the best way to track expressions is with `pg_get_partkeydef`. Thoughts?

Reply:
@uhoh-itsmaciek Mh, I see - yeah, that looks a bit complicated to extract correctly.

I think doing two fields, one repeated field for column numbers (as you had here initially), and then an additional `partitioned_by` field would be useful.

I'm particularly thinking for the case where we have a list of columns on the table, and want to highlight which ones of them are included in the partition key - for that its useful to have the list of columns as numbers, instead of only the SQL that describes the partition key.